app/validators/questionnaire_schema.py

```diff
@@ -0,0 +1,298 @@
+
+                string_path = ".".join(block_path)
+                # Since unique_ids_per_block is a set, duplicate ids will only be recorded once within the block.
+                unique_ids_per_block[string_path].add(value)
+            else:
+                non_block_ids.append(value)
+
+        for block_ids in unique_ids_per_block.values():
+            all_ids.extend(block_ids)
+        all_ids.extend(non_block_ids)
+
+        return all_ids
+
+    @cached_property
+    def id_paths(self):
+        """
+        These values will be returned with the json path to them through the object e.g.
+            - 'sections.[0].groups[0].blocks[1].question_variants[0].question.question-2'
+
+        Returns: generator yielding (path, value) tuples
+        """
+        ignored_paths = [
+            "routing_rules",
+            "skip_conditions",
+            "when",
+            "edit_block.question.answers",
+            "add_block.question.answers",
+            "remove_block.question.answers",
+            "edit_block.question_variants.answers",
+            "add_block.question_variants.answers",
+            "remove_block.question_variants.answers",
+        ]
+
+        for match in parse("$..id").find(self.schema):
+            full_path = str(match.full_path)
+            if not any(ignored_path in full_path for ignored_path in ignored_paths):
+                yield str(match.full_path)[:-3], match.value
+
+    @cached_property
+    def answer_id_to_option_values_map(self):
+        answer_id_to_option_values_map = defaultdict(set)
+
+        for answer in self.answers:
+            if "options" not in answer:
+                continue
+
+            answer_id = answer["id"]
+            option_values = [option["value"] for option in answer["options"]]
+
+            answer_id_to_option_values_map[answer_id].update(option_values)
+
+        return answer_id_to_option_values_map
+
+    @cached_property
+    def answers(self):
+        for question, _ in self.questions_with_context:
+            for answer in question["answers"]:
+                yield answer
+
+    @lru_cache
+    def get_answer(self, answer_id):
+        return self.answers_with_context[answer_id]["answer"]
+
+    @lru_cache
+    def get_group(self, group_id):
+        return self.groups_by_id[group_id]
+
+    @lru_cache
+    def get_section(self, section_id):
+        return self.sections_by_id[section_id]
+
+    @lru_cache
+    def get_block(self, block_id):
+        return self.blocks_by_id[block_id]
+
+    @lru_cache
+    def get_blocks(self, **filters):
+        conditions = []
+        for key, value in filters.items():
+            conditions.append(f'@.{key}=="{value}"')
+
+        if conditions:
+            final_condition = " & ".join(conditions)
+            return jp.match(f"$..blocks[?({final_condition})]", self.schema)
+        return self.blocks
+
+    @lru_cache
+    def get_block_key_context(self, block_id, key_name):
```

Comment:
This name confused me a little. I think something like `get_object_containing_key` would be clearer. It might also be worth changing `block_id` for a variable that contains some schema, as the two places that call this both already have the required schema.